File: array/MonotonicStack/python/MonotonicStack.py
def next_greater_elements(values: list[int]) -> list[int]:
    answer = [-1] * len(values)
    stack: list[int] = []

    for index in range(len(values) - 1, -1, -1):
        while stack and stack[-1] <= values[index]:
            stack.pop()
        if stack:
            answer[index] = stack[-1]
        stack.append(values[index])

    return answer

File: array/MonotonicStack/rust/src/monotonic_stack.rs
pub fn next_greater_elements(values: &[i64]) -> Vec<i64> {
    let mut answer = vec![-1_i64; values.len()];
    let mut stack: Vec<i64> = Vec::with_capacity(values.len());

    for index in (0..values.len()).rev() {
        while let Some(&top) = stack.last() {
            if top <= values[index] {
                stack.pop();
            } else {
                break;
            }
        }
        if let Some(&top) = stack.last() {
            answer[index] = top;
        }
        stack.push(values[index]);
    }

    answer
}
